### PR TITLE
New rule: FM recommendations

### LINKED ISSUE #1187: new rule, function module recommendations
much like https://docs.abapopenchecks.org/checks/53/

### PR DESCRIPTION
**Pull request by g-back**
closes #1187

### CODE REVIEW

File: packages/core/src/rules/function_module_recommendations.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  private conf = new FunctionModuleRecommendationsConf();
+
+  public getMetadata(): IRuleMetadata {
+    return {
+      key: "function_module_recommendations",
+      title: "Function Module Recommendations",
+      shortDescription: `Function Module Recommendations`,
+      extendedInformation: `https://docs.abapopenchecks.org/checks/53/`,
+      tags: [RuleTag.Styleguide],
```
Comment (larshp) [inline packages/core/src/rules/function_module_recommendations.ts:47]: If its part of the ABAP style guide somewhere, please add a link under extendedInformation

File: packages/core/src/rules/function_module_recommendations.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+      // only check constant FM names
+      if (!funcName.startsWith("'")) {
+        continue;
+      }
+      // remove leading and trailing single quote
+      funcName = funcName.slice(1, funcName.length - 1);
+      const index = this.conf.recommendations.findIndex(
+        i => i.name.toUpperCase() === funcName && (i.from === undefined || configVersion >= i.from));
+      if (index >= 0) {
+        issues.push(Issue.atToken(file, token, this.getMessage(index), this.getMetadata().key));
```
Comment (larshp) [inline packages/core/src/rules/function_module_recommendations.ts:78]: add severity, example: https://github.com/abaplint/abaplint/blob/198643a85378319ee368049878d182cb7a05a20c/packages/core/src/rules/begin_single_include.ts#L61

**Review by larshp — CHANGES_REQUESTED** [review]

**Comment by FreHu** [comment]
I'd also recommend the FM `function_exists`, it should be replaced by calling the function in a try-catch instead of checking for existence before calling. This approach is much faster assuming the function exists.

**Review by larshp — APPROVED** [review]

**Comment by larshp** [comment]
thanks 👍

**Pull request merged**